test/MatrixFields/matrix_field_broadcasting.jl

```diff
@@ -0,0 +1,816 @@
+using Test
+using JET
+using Random: seed!
+using LinearAlgebra: I, mul!
+using BandedMatrices: band
+
+using ClimaCore.MatrixFields
+import ClimaCore:
+    Geometry, Domains, Meshes, Topologies, Hypsography, Spaces, Fields
+import ClimaComms
+
+# Using @benchmark from BenchmarkTools is extremely slow; it appears to keep
+# triggering recompilations and allocating a lot of memory in the process.
+# This macro returns the minimum time (in seconds) required to run the
+# expression after it has been compiled.
+macro benchmark(expression)
+    return quote
+        $(esc(expression)) # Compile the expression first. Use esc for hygiene.
+        best_time = Inf
+        start_time = time_ns()
+        while time_ns() - start_time < 1e8 # Benchmark for 0.1 s (1e8 ns).
+            best_time = min(best_time, @elapsed $(esc(expression)))
+        end
+        best_time
+    end
+end
+
+# This has to be its own function in order to correctly measure its allocations.
+call_array_func(
+    ref_set_result!::F,
+    ref_result_arrays,
+    inputs_arrays,
+    temp_values_arrays,
+) where {F} = foreach(
+    ref_set_result!,
+    ref_result_arrays,
+    inputs_arrays...,
+    temp_values_arrays...,
+)
+
+function test_matrix_broadcast_against_array_reference(;
+    test_name,
+    inputs,
+    get_result::F1,
+    set_result!::F2,
+    temp_values = (),
+    ref_set_result!::F3,
+    time_ratio_limit = 1,
+    max_error_limit = 3,
+) where {F1, F2, F3}
+    @testset "$test_name" begin
+        result = get_result(inputs...)
+
+        # Fill all output fields with NaNs for testing correctness.
+        result .*= NaN
+        for temp_value in temp_values
+            temp_value .*= NaN
+        end
+
+        ref_result_arrays = MatrixFields.field2arrays(result)
+        inputs_arrays = map(MatrixFields.field2arrays, inputs)
+        temp_values_arrays = map(MatrixFields.field2arrays, temp_values)
+
+        best_time = @benchmark set_result!(result, inputs...)
+        best_ref_time = @benchmark call_array_func(
+            ref_set_result!,
+            ref_result_arrays,
+            inputs_arrays,
+            temp_values_arrays,
+        )
+
+        # Compute the maximum error as an integer multiple of machine epsilon.
+        result_arrays = MatrixFields.field2arrays(result)
+        max_error =
+            maximum(zip(result_arrays, ref_result_arrays)) do (array, ref_array)
+                maximum(zip(array, ref_array)) do (value, ref_value)
+                    Int(abs(value - ref_value) / eps(ref_value))
+                end
+            end
+
+        @info "$test_name:\n\tBest Time = $best_time s\n\tBest Reference Time \
+               = $best_ref_time s\n\tMaximum Error = $max_error eps"
+
+        # Test that set_result! is performant compared to ref_set_result!.
+        @test best_time < time_ratio_limit * best_ref_time
```

Comment:
@dennisYatunin, we don't need to use a test here-- these tests are unfortunately pretty flakey. I'm fine with changing this to be an `@info` print to the log of the results. Keep the correctness, allocation, and JET tests though!

Reply:
I think I've actually been getting fairly consistent timing results, since I'm taking the minimum time from a large number of trials. There's a big difference between running locally and on Buildkite, but that's probably just because the cluster can do vectorized operations more quickly than my computer. If you want, though, I can just comment out this test; the info statement above it prints all the relevant results.

Reply:
Okay, either way is fine with me-- if you leave the test in, just include a buffer that we can bump to help alleviate flakiness from timing variance.